From ae83b3659f5fa14c25d5e554cf2b35e15c7bbb7c Mon Sep 17 00:00:00 2001
From: "Du, Changbin" <changbin.du@intel.com>
Date: Thu, 15 Jan 2015 20:10:24 +0800
Subject: [PATCH 1/1] usb: hub.c: do not clear BOS field during reset device

The old BOS descriptor may still be used before allocating a new bos.
So we cannot set the udev->bos to NULL before that, just keep what it
was. It will be overwrite when allocating a new one.

Crash log:
BUG: unable to handle kernel NULL pointer dereference at 0000000000000010
IP: [<ffffffff8171f98d>] usb_enable_link_state+0x2d/0x2f0
Call Trace:
[<ffffffff8171ed5b>] ? usb_set_lpm_timeout+0x12b/0x140
[<ffffffff8171fcd1>] usb_enable_lpm+0x81/0xa0
[<ffffffff8171fdd8>] usb_disable_lpm+0xa8/0xc0
[<ffffffff8171fe1c>] usb_unlocked_disable_lpm+0x2c/0x50
[<ffffffff81723933>] usb_reset_and_verify_device+0xc3/0x710
[<ffffffff8172c4ed>] ? usb_sg_wait+0x13d/0x190
[<ffffffff81724743>] usb_reset_device+0x133/0x280
[<ffffffff8179ccd1>] usb_stor_port_reset+0x61/0x70
[<ffffffff8179cd68>] usb_stor_invoke_transport+0x88/0x520

Change-Id: Ifa1214b01906db0b5a68152d2948f0ed1b512174
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-16691
Signed-off-by: Du, Changbin <changbin.du@intel.com>
---
 drivers/usb/core/hub.c |    7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

Index: b/drivers/usb/core/hub.c
===================================================================
--- a/drivers/usb/core/hub.c	2016-03-16 10:07:48.793943603 +0800
+++ b/drivers/usb/core/hub.c	2016-03-16 10:21:54.407260011 +0800
@@ -5486,8 +5486,11 @@
 re_enumerate:
 	/* LPM state doesn't matter when we're about to destroy the device. */
 	hub_port_logical_disconnect(parent_hub, port1);
-	usb_release_bos_descriptor(udev);
-	udev->bos = bos;
+	/* release the new BOS descriptor allocated  by hub_port_init() */
+	if (udev->bos != bos) {
+		usb_release_bos_descriptor(udev);
+		udev->bos = bos;
+	}
 	return -ENODEV;
 }
 
